# Number Theory


## Preface


This repository concerns the textbook <u>Introduction to Analytic Number Theory</u> 
by Tom Apostol. This introductory Python notebook lays out my motivation, my approach 
to solving the problems, comments on format and so on.


$LaTeX$ formatting is something of a labor; so the $LaTeX \; patterns$ notebook 
contains notes on the accumulation I needed to sort out. 


## Motivation


### On the unreasonable effectiveness of asking for help


This collection of notebooks takes up the study of number theory, 
inspired by two people: Caltech mathematics professor Tom Apostol
and Scientific American columnist Martin Gardner. Further 
inspiration was provided by Paul Erdös and Ross Honsberger. 



Here is a story I made up:


> A student of rock climbing was led by his instructor to a rock 
> formation one day. 'Here is a problem: Climb to the top' the instructor
> said, pointing to a vertical cliff face. Whereupon (the instructor
> having departed) the student examined the cliff face and found it
> to be vertical, smooth and featureless. There was not so much as a
> tiny bump of a handhold. The student sat upon the ground, staring at
> the cliff. A year later he returned to his instructor. 'I have not
> solved the problem yet' he announced in a despondent voice.
> 'Oh?...' his instructor replied '...I assumed you had died.'


Some years ago--never mind how long precisely--having little or no money in my 
purse, and nothing particular to interest me in my home town, I thought I would 
travel about a little and see the learned part of the world. In due course I arrived 
at the California Institute of Technology where I had the privilege of attending 
Professor Apostol's lectures on calculus and related topics. In keeping with the 
Institute's undergraduate culture of irreverence, my fellow classmates and I referred 
to Tom Apostol as '*Tommy*'. I see him in my mind's eye standing before the blackboard 
in Noyes Laboratory, all business delivering a lecture on a version of calculus 
much more *carefully explained* relative to what we had seen in highschool. This was
our introduction to mathematical rigor. At certain points in the narrative he would pause
for dramatic effect and make an impossible, outrageous mathematical claim. There would 
follow another pause to let the claim register... we might scoff 'oh please' but 
only quietly to ourselves... and then Professor 
Apostol would continue with a portentous remark, "Now nobody can stop me from *making 
this claim*...". He was daring us to guess how it might in fact be true. 
And then his logical argument would follow step by step, at the end of which we were
convinced: The ridiculous claim was quite reasonable. It became part of what was true. 


As Freshmen we did not always appreciate the fun in this narrative style. 


In retrospect I appreciate Tommy's lecture style now. He took the burden of proof
very seriously, never took the easy way out typified by the phrase 'it turns out that...' 
Plus his pace of delivery always gave one a sense of forging ahead.


My intent here is to circle back and start over with Tommy on the topic of analytic 
number theory. Working with prime numbers and composites begins in grade school, 
accompanied by some confusion over the status of $1$. The first chapter of Tommy's book
**Introduction to Analytic Number Theory** (abbreviated ANT here) provides a brief
Historical Introduction before charging forward with non-intuitive building blocks
centered on the greatest common divisor. Chapter two introduces the Möbius $\mu(n)$ 
function and Euler's totient $\varphi(n)$; and then Dirichlet multiplication. 
These two chapters are my price of admission. They are the smooth rock face one 
climbs to reach the next plateau of topics: The distribution of primes, finite Abelian 
groups, quadratic reciprocity, ... and as early as chapter 3 the 
mysterious Riemann zeta function.  


I refer to the metaphor of the student rock climber. He is intended to represent the
fool I learn from. I have no pretention of mathematical talent but I have spent enough 
time trying to solve math problems to conclude that an honest try is worth a shot. 
If that effort reaches a dead end: Getting help is in no sense a failure. To this end, 
two persons have preceded me in writing up public solutions to ANT problems: One Sean 
Li and the estimable Greg (Chip) Hurst. I rely primarily on Chip's remarkable ANT 
solution set available for download online in *pdf* format.


Each chapter gets up to three notebooks: Narrative, Examples and Solutions.  The
supporting Python code accumulates in a module file `ant.py`. As noted above there
is a notebook on `LaTeX patterns`. There is finally a "Where are you headed?"
notebook called `QuoVadis`. These are notes on the big picture but I'm afraid they
are not terribly illuminating just yet. 


For a given problem my method is some variant of this procedure:


- Write the problem in the Jupyter notebook
- Work some examples
- Write down an obvious solution should one appear at this point
- If not: Proceed to write some test code in the examples notebook
    - Expand the `ant.py` module if necessary
- Refer to the text, try out some ideas
- Write down an obvious solution should one appear at this point
- If not: Acknowledge that help is needed
    - Look at Chip's solution... this takes some serious concentration
    - Write up a (usually more *elaborate*) version of Chip's solution


My aesthetic is to strive for readability: I should be able to follow a
solution without needing pen and paper.



### further motivation


Going back further in my personal history: Early on I discovered a dichotomy as 
I struggled to make sense of the world. To begin with, the book **Dinosaurs and More Dinosaurs** 
(1971) made it clear that there really are -- or were -- monsters
who did not know or care about me. Their existence indicated that the world really was more fantastical 
than anything I could imagine. Dinosaurs were a first clue. I also found two books 
written about a girl named Alice. Here the matter was clearly fiction but
nevertheless I found it to be a world that I wanted to *visit*. I became lost in the
story until it ended; and upon setting it down I found myself back in the ordinary world. 
All this suggests a dichotomy: Is the actual, apparent world we inhabit ordinary or extraordinary? 
Dinosaurs and Red Queens are real when the book is open... but what happens when the book closes? 
Do they persist? I responded in what seemed the only logical way: I looked for more books.



In middle school I discovered computers and Martin Gardner. The latter was at first merely
the name on the by-line of the recurring *Mathematical Games* feature in Scientific American. 
Whoever he was, Mr. Gardner invited me to climb up onto the mantle above the fireplace and peer
into the Looking Glass room. Then as the mirror dissolved in a mist I was invited to 
pass through and climb down into the Looking Glass house. Here ideas mattered in proportion 
to their strangeness. My bemused tour guide spoke of logic puzzles, of paradoxes, of prisoners 
both clever and unfortunate, of inwardly exploding fractals, of planar graphs, of colored maps, 
of games designed around a search for truth, and on and on. There was practical advice on how 
to function in four dimensions and there was a peculiar pencil-and-paper game passed along
by someone named Conway.


- I would like to present the backbone of yes / no / unknown as what made the Gardner world finally real
    - Dinosaurs are real but long gone... unless we care to turn our focus to birds
    - Fiction makes no pretensions of being actual
    - Mathematics suffers the burden of being provable
        - But still exceeds anything we might find in a dream or a drawing



Having arrived at this state of being, 'there is a big world to discover', 
the question evolves into 'How then shall we explore?' The fuse is lit, 
we are off to explore. But what happens when the trail grows cold, when
we hit a wall, run out of ideas to try? This is the point of the story
of the rock climbing student. Consider an indirect method where one develops 
a skill called 'getting help'.  
Sometimes help is not available. Sometimes sources of 
help are hidden. Sometimes they walk away and we have to go chase them
down. Of course the reverse can be true as well: We can become so
habituated to help that we become fearful of the problem. So there is 
a balance to strike; but I feel that it is too easy to err on the 
side of not finding that help. 


This goes higher up.


"If anything has kept me at Caltech all these years, it has been the 
quality of the undergraduate student body." -Tom Apostol


This was Professor Apostol paying his students the 
highest compliment in an interview he gave in 1997 
(Donald J. Albers (1997) An Interview with Tom Apostol, The College 
Mathematics Journal, 28:4, 250-270, DOI: 10.1080/07468342.1997.11973873). 
The interview is a worthwhile read and I feel it reinforces Martin
Gardner's implicit contention in his work that there is something out 
there we can touch, given pencils, persistence and enough scratch paper.



>This comment belongs in the section below on solution style: 
On starting to work through **Introduction to Analytic Number Theory**
I found I understood theorems and proofs on the 'in the moment' sense;
but internalizing them proves challenging.




Thanks Tommy, thanks Martin. Thanks Chip. Thanks Paul, and thank you Ross.
And thanks to many others.

In [2]:
# An example calculation
# From Mathematical Games in the July 1976 issue of Scientific American: What is the sum of 
# positive-valued fractions whose numerators are the odd numbers 1, 3, 5, 7, ... and whose 
# denominators are the powers of two 1, 2, 4, 8, ...  MG suggests this might tot up to 6.

n_fractions = 30
numerator = 1
denominator = 1
gardner_sum = 0
for i in range(n_fractions):
    gardner_sum += numerator / denominator
    numerator += 2
    denominator *= 2
    print(round(gardner_sum, 6))
print('End result: ', gardner_sum)

1.0
2.5
3.75
4.625
5.1875
5.53125
5.734375
5.851562
5.917969
5.955078
5.975586
5.986816
5.99292
5.996216
5.997986
5.998932
5.999435
5.999702
5.999844
5.999918
5.999957
5.999978
5.999988
5.999994
5.999997
5.999998
5.999999
6.0
6.0
6.0
End result:  5.999999882653356


### Solution style elements


Learning number theory depends on having time, cash and patience. My work here is a combination of solutions
and narrative. The latter is intended to help articulate and internalize the tools and ideas; placed as separated
text between solutions. For example the Mobius function is presented as something of a *fait accompli* so I
devote some writing to *why* it is defined as it is.



Another example intended as slightly mystifying: Suppose for two positive integers $a$ and $b$ we have that their
greatest common divisor is $1$: $(a, b) = 1$. Then there exist integers $x$ and $y$ such that $c = c \cdot (ax + by)$.


A telescope example: $\alpha^n-1$ suggests telescoping factorization via factor $\alpha \; - \; 1$. A step 
further: $\alpha^n + 1$ might be factored via alternating signs $+ \; - \; + \; - \; \dots - \; +$. 
(See problems **1.16** and **1.17**.)


On *ducking the wave*, i.e. missing an elegant solution in favor of more pedestrian methods. I will plead no contest 
as I may start working a problem from cases; which suggests *mechanical* solutions that duck the wave. 
(See problem 2.5)


My notion of adequate solutions and proofs is not terribly rigorous. I think the criterion is 'have I convinced
myself?' (and try to be consistent about what 'convincing' means).

I'm using markdown and LaTeX for equations. What is needed here is a nexus reference to my typesetting notes.


This raises the question of how to communicate "this completes the proof"
Various tombstones are available via Unicode: the filled tombstone: &#x220e; the outline: &#x1f78e;
a large thin-line square: &#x2b1c; a medium square: &#x25fb;
and what I will use, a heavier-line square.
Example: ```$(c, d)=1.\;\;$ &#x2610;``` renders as $(c, d)=1.\;\;$ &#x2610;

## **The Prime Factorization Theorem** 


The PFT Theorem: *Any positive integer greater than 1 has a unique prime factorization.*


This proof presumes the theorem to be *false* and arrives at a contradiction.


To clarify one point: A *unique* prime factorization $F(N > 1)$ is a set of prime numbers $\{ p_i \}$ such that 
$\prod p_i = N$. This set is order-independent and prime factors may appear in the set more 
than once, as in $F(36) = \lbrace 3, 2, 2, 3 \rbrace$.


Suppose the theorem is false: There are positive integers with two or more distinct prime factorizations.
Then there is a smallest such number $S$ such that any number less than $S$ has but one prime factorization.
Such an $S$ is certainly greater than $7$.


$S$ can be factored in two ways: $S = \alpha \cdot A$ and $S = \beta \cdot B$. Here $\alpha$ and $\beta$
are distinct prime factors of $S$ so $\alpha \ne \beta$. $A$ and $B$ are the residual factors. They 
might be be prime or composite. 


For definiteness let us take $\alpha < \beta$ implying $A > B$.

In $S = \alpha \cdot A = \beta \cdot B$ all five values must be odd. 
Since $\alpha < \beta$ and $B < A$ we infer $A - B$ and $\beta - \alpha$ are positive even numbers. 
Hence $\beta$ is equal to $\alpha$ plus an even number that we can parameterize as $2r$: 
$\beta = \alpha + 2r$ with $r \ge 1$.


In like fashion $A = B + 2q$ with $q \ge 1$: Difference parameters $q$ and $r$ are both positive integers.


Substituting we have $S = \alpha \cdot (B + 2 q) = (\alpha + 2 r) \cdot B$. 
Hence $\alpha B + 2 q \alpha = \alpha B + 2 r B$ so that $q \alpha = r B$. 



This number $q \alpha$ also known as $r B$ is less than $S$ per the expressions for $S$ above.



Since $q \alpha$ is less than $S$ it has only one prime factorization. 
Since $\alpha$ is not one of the factors in the prime factorization of $B$ it must be a factor of $r$ so we can introduce 
another new parameter $m$: $r = m \alpha$ for some $m > 0$.


Returning to the expression $S = \beta B$ we now have the substitutions:


$\begin{align}S = \beta B = (\alpha + 2 r) B = (\alpha + 2 m \alpha) B = \alpha (1 + 2 m) B\end{align}$


This means that $\beta B = \alpha (1 + 2m) B$ where $(1 + 2m)$ is greater than one. That is, $\beta = \alpha (1 + 2m)$.
This violates the presumption that $\alpha$ does not divide $\beta$. In fact $\beta$ is defined as a prime number
but is now shown to be composite. This is the contradiction in following the presumption of two unique prime 
factorizations of $S$. 


To conclude: Supposing an $S$ with more than one unique factorization has led to a contradiction.   &#x2610;